python/lsst/meas/base/forcedPhotCcd.py

```diff
@@ -35,34 +37,107 @@
-    from the set of coadds which intersect with the calexp.  a set of reference catalog
-    from a coadd which overlaps it.  It needs the calexp id (e.g. visit, raft, sensor), but it
-    also uses the tract to decide what set of coadds to use.  The references from the tract
-    whose patches intersect with the calexp are used.
+    Required because we need to add "tract" to the raw data ID keys (defined as whatever we
+    use for 'src') when no tract is provided (so that the user is not required to know
+    which tracts are spanned by the raw data ID).
+
+    This IdContainer assumes that a calexp is being measured using the detection information,
+    a set of reference catalogs, from the set of coadds which intersect with the calexp.
+    It needs the calexp id (e.g. visit, raft, sensor), but is also uses the tract to decide
+    what set of coadds to use.  The references from the tract whose patches intersect with
+    the calexp are used.
     """
+    def _addDataRef(self, namespace, dataId, tract):
+        """Construct a dataRef based on dataId, but with an added tract key"""
+        forcedDataId = dataId.copy()
+        forcedDataId['tract'] = tract
+        dataRef = namespace.butler.dataRef(datasetType=self.datasetType, dataId=forcedDataId)
+        self.refList.append(dataRef)
+
     def makeDataRefList(self, namespace):
         """Make self.refList from self.idList
         """
+        if self.datasetType is None:
+            raise RuntimeError("Must call setDatasetType first")
+        log = lsst.pex.logging.Log.getDefaultLog()
+        skymap = None
+        visitTract = {}  # Set of tracts for each visit
+        visitRefs = {}   # List of data references for each visit
         for dataId in self.idList:
             if "tract" not in dataId:
-                raise argparse.ArgumentError(None, "--id must include tract")
-            tract = dataId.pop("tract")
-            # making a DataRef for src fills out any missing keys and allows us to iterate
-            for srcDataRef in namespace.butler.subset("src", dataId=dataId):
-                forcedDataId = srcDataRef.dataId.copy()
-                forcedDataId['tract'] = tract
-                dataRef = namespace.butler.dataRef(
-                    datasetType = "forced_src",
-                    dataId = forcedDataId,
-                    )
-                self.refList.append(dataRef)
+                # Discover which tracts the data overlaps
+                log.info("Reading WCS for components of dataId=%s to determine tracts" % (dict(dataId),))
+                if skymap is None:
+                    skymap = namespace.butler.get(namespace.config.coaddName + "Coadd_skyMap")
+
+                for ref in namespace.butler.subset("calexp", dataId=dataId):
+                    if not ref.datasetExists("calexp"):
+                        continue
+
+                    visit = ref.dataId["visit"]
+                    visitRefs = collections.defaultdict(list)
```

Comment:
This looks wrong: every time around the loop you're replacing whatever `visitRefs` used to refer to with a new, empty `defaultdict`. You probably want to move this line outside the loop, replacing line 71.